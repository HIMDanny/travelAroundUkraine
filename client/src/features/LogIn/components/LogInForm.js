/* eslint-disable no-unused-vars */
import React, { useState } from 'react';
import { Box, Button, Typography, CircularProgress, Grid } from '@mui/material';
import { useNavigate } from 'react-router-dom';
import { TextField } from 'formik-mui';
import { Formik, Field, Form } from 'formik';
import * as yup from 'yup';
import { useDispatch, useSelector } from 'react-redux';
import postLogIn from '../../../api/postLogIn';
import { setIsLogin } from '../../../store/slices/userSlice/userSlice';
import { migrateCart, fetchCart } from '../../../store/slices/cartSlice/cartSlice';

const buttonSX = {
  width: { xs: '86px', mobile: '110px', tablet: '140px', laptop: '150px' },
  px: 0,
  py: { xs: '8px', mobile: '12px', desktop: '14px' },
  mx: 4,
  fontSize: { xs: '12px', laptop: '14px' },
};

const preloaderSX = {
  my: 5,
  width: '80px !important',
  height: '80pX !important',
};

const gridItemSx = {
  width: { xs: '80%', tablet: '60%', laptop: '40%' },
  my: { xs: '4%', tablet: '3%' },
};

const LogInForm = () => {
  const [errorMassage, setErrorMassage] = useState(null);
  const [isLoading, setIsLoading] = useState(false);
  const cart = useSelector((store) => store.cart.data);

  const navigate = useNavigate();
  const dispatch = useDispatch();

  const myHandleSubmit = async (values) => {
    setIsLoading(true);

    const response = await postLogIn(values).then((res) => res);

    if (response.success) {
      localStorage.setItem('token', response.token);
      setErrorMassage(null);
      navigate('/');
      dispatch(setIsLogin(true));

      dispatch(fetchCart())
        .unwrap()
        .then(() => {
          const localCart = JSON.parse(localStorage.getItem('cart'));
          localCart && dispatch(migrateCart(localCart));
        });
    } else {
      setErrorMassage(response?.password || response?.loginOrEmail);
    }

    setIsLoading(false);
  };

  const handleClickSignUp = () => {
    navigate('/signup');
  };

  const initialValues = {
    loginOrEmail: '',
    password: '',
  };

  const validationSchema = yup.object().shape({
    loginOrEmail: yup.string().required('Email or login are required'),
    password: yup.string().min(7, 'Password must have at least 7 characters').required('Password is required!'),
  });

  return (
    <>
      {isLoading ? (
        <CircularProgress sx={preloaderSX} />
      ) : (
        <Formik initialValues={initialValues} validationSchema={validationSchema} onSubmit={myHandleSubmit}>
          {() => (
            <Form>
              <Grid container direction="column" alignItems="center">
                <Grid item sx={gridItemSx}>
                  <Field component={TextField} name="loginOrEmail" label="Email" fullWidth />
                </Grid>
                <Grid sx={gridItemSx}>
                  <Field component={TextField} type="password" name="password" label="Password" fullWidth />
                </Grid>
              </Grid>
              {errorMassage && (
                <Typography sx={{ color: '#d32f2f', textAlign: 'center' }}>Password or login is not correct</Typography>
              )}
              <Box sx={{ my: { xs: 1.5, laptop: 2.5 } }}>
                <Button sx={buttonSX} onClick={handleClickSignUp}>
                  sign up
                </Button>
                <Button type="submit" sx={buttonSX}>
                  log in
                </Button>
              </Box>
            </Form>
          )}
        </Formik>
      )}
      {}
    </>
  );
};
export default LogInForm;
