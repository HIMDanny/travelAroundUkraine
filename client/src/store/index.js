import { configureStore } from '@reduxjs/toolkit';
import userReducer from './slices/userSlice/userSlice';
import tourReducer from './slices/tourSlice/tourSlice';
import catalogueReducer from './slices/catalogueSlice/catalogueSlice';
import cartReducer from './slices/cartSlice/cartSlice';
import orderSlice from './slices/orderSlice/orderSlice';
import inFavoritesReducer from './slices/inFavoritesSlice/inFavoritesSlice';

const store = configureStore({
  reducer: {
    userReducer,
    tour: tourReducer,
    catalogue: catalogueReducer,
    cart: cartReducer,
    order: orderSlice,
    favorites: inFavoritesReducer,
  },
});

export default store;
