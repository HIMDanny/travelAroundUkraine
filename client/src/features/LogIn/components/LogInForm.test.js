import LogInForm from './LogInForm';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../../../store';
import { render } from '@testing-library/react';

jest.mock('@mui/material/Box', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Button', () => ({ children }) => <button>{children}</button>);
jest.mock('@mui/material/Typography', () => ({ children }) => <p>{children}</p>);
jest.mock('../../../api/postLogIn', () => () => () => {});

describe('Render LogInForm', () => {
  test('should LogInForm render', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <LogInForm />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
