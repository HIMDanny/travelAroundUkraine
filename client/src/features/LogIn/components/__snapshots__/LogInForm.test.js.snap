// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render LogInForm should LogInForm render 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-t0zib5-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-93lq6l-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r0:"
            id=":r0:-label"
          >
            Email
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              name="loginOrEmail"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Email
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root css-7net31-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r1:"
            id=":r1:-label"
          >
            Password
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r1:"
              name="password"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Password
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
    <div>
      <button>
        sign up
      </button>
      <button>
        log in
      </button>
    </div>
  </form>
</DocumentFragment>
`;
